File: nodejs/cpanel.js

var net = require('net'); 
var Q = require('q');
var xml2js = require('xml2js');

var liveApi = function() {
	var self = this;
	this.isConnected = false;
	this.sockfile = process.env.CPANEL_PHPCONNECT_SOCKET;
	this.queue = [];
	console.log("init");
	this.def = Q.defer();
	this.def.resolve(true);

};


liveApi.prototype.end = function(callback) {
	var self = this;
	console.log("end called");
	self.isConnected = false;
	var deferred = Q.defer();
	deferred.promise.nodeify(callback);
		self.send('<cpanelxml shutdown="1" />').then(function(){
			self.client.once('data', function(data) { 
					deferred.resolve(true);
					process.exit(0);
				});
			
		});
	
    return deferred.promise;

}
liveApi.prototype.connect = function(callback) {
	console.log("connect called");
	var self = this;
	var deferred = Q.defer();
	deferred.promise.nodeify(callback);
	if (!self.isConnected)
	{
		self.client = net.createConnection({path: self.sockfile, allowHalfOpen: false}, function() {
			console.log("connected");
			self.client.setNoDelay(true);
			self.isConnected = true;
		
			self.send('<cpaneljson enable="1">').then(function(){
				console.log("setup data sent");
				self.client.once('data', function(data) { 
					deferred.resolve(true);
				});
				
			});
		});
	} else {
		deferred.resolve(true);
	}
	
	
    return deferred.promise;
};
liveApi.prototype.send = function(code, callback) {
	var self = this;
	var deferred = Q.defer();
	deferred.promise.nodeify(callback);
	self.client.write((code.length) + "\n" + code, function() {
		deferred.resolve(true);
	});
	return deferred.promise;
};


liveApi.prototype.execWork = function(code, callback) {
	var self = this;
	console.log("exec called: " + code);
	
	var deferred = Q.defer();
	deferred.promise.nodeify(callback);

	self.connect().then(function() {

	self.client.once('data', function(data) {
			console.log("rec data callback:" + data.toString());
			var parser = new xml2js.Parser({charsAsChildren: true, ignoreAttrs: true, charkey: 'data'});
			parser.parseString(data, function (err, result) {
				if (err != undefined)
				{
					console.log("parse fail");

					deferred.reject(err);
					

				} else {
					console.log("parse ok");
					var ob = result.cpanelresult;
					if(typeof(ob) == "string")
					{
						ob = JSON.parse(result.cpanelresult);
					}
					
					if (ob.cpanelresult != undefined)
					{
						ob = ob.cpanelresult;
					}
					if(ob.error != undefined && ob.data != undefined)
					{
						ob.data = JSON.parse(ob.data);
						if(ob.data.cpanelresult != undefined)
						{
							ob = ob.data.cpanelresult;
						}
					}
					
					deferred.resolve(ob);
					
				}
			});

		});

		self.client.write((code.length) + "\n" + code, function() {
		console.log("wrote data: "+ code);
		

	});
});


	return deferred.promise;
};


liveApi.prototype.run = function() {
	var self = this;
	
	var next = self.queue.pop();
			if(next)
			{
						
					
			var deferred = next[1];
			var code = next[0];
			var task = Q.defer();
			self.def = task;
			self.execWork(code).then(function(data) {
					deferred.resolve(data);
					self.run();
				});
		} else {
			self.def.resolve(true);
		}
};
liveApi.prototype.exec = function(code) {
	var self = this;
	console.log("Exec request: " + code);
	var deferred = Q.defer();

	if (self.def.promise.isFulfilled())
	{
		
		console.log("no que");
		self.queue.push([code, deferred]);
		self.run();
		
	} else {
		console.log("que");
		self.queue.push([code, deferred]);
	}
	
	return deferred.promise;
	
};


liveApi.prototype.fetch = function(input, callback) {

	var self = this;
	var deferred = Q.defer();
	deferred.promise.nodeify(callback);
	var query = self.exec('<cpanel print="' + input + '">');
	deferred.resolve(query);
	return deferred.promise;
};

liveApi.prototype.api = function(reqtype, version, module, func, apiargs, callback)
{
	var self = this;
	var deferred = Q.defer();
	
	
	var input = {"module": module, "reqtype" : reqtype, "func": func, "apiversion":version};
	
	if (Array.isArray(apiargs)) {
		input.args = apiargs;
	}
	
	var code = "<cpanelaction>\n" + JSON.stringify(input) + "\n</cpanelaction>";
	
	
	var query = self.exec(code);
	deferred.resolve(query);
	return deferred.promise;
}

liveApi.prototype.api1 = function(module, func, apiargs, callback) {

	var self = this;
	var deferred = Q.defer();
	deferred.promise.nodeify(callback);
	var query = self.api('exec', 1, module, func, apiargs);
	deferred.resolve(query);
	return deferred.promise;
};

liveApi.prototype.api2 = function(module, func, apiargs, callback) {

	var self = this;
	var deferred = Q.defer();
	deferred.promise.nodeify(callback);
	var query = self.api('exec', 2, module, func, apiargs);
	deferred.resolve(query);
	return deferred.promise;
};

liveApi.prototype.api3 = function(module, func, apiargs, callback) {

	var self = this;
	var deferred = Q.defer();
	deferred.promise.nodeify(callback);
	var query = self.api('exec', 3, module, func, apiargs);
	deferred.resolve(query);
	return deferred.promise;
};

liveApi.prototype.uapi = function(module, func, apiargs, callback) {

	var self = this;
	var deferred = Q.defer();
	deferred.promise.nodeify(callback);
	var query = self.api('exec', 'uapi', module, func, apiargs);
	deferred.resolve(query);
	return deferred.promise;
};






module.exports = liveApi;

File: python/cpanel.py
import sys, os, socket, json, pprint
class LiveAPI(object):
    """Cpanel Live api"""
    connected = False
    s = None
    _result = None

    def __init__(self):
        self.connected = True
        sockfile = os.environ['CPANEL_CONNECT_SOCKET']
        if not sockfile:
            raise RuntimeError('There was a problem fetching the env variable containing the path to the socket')
        try:
            self.s = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        except socket.error, msg:
            self.s = None
        try:
            self.s.connect(sockfile)
        except socket.error, msg:
            self.connected = False
            raise
        self.s.setblocking(1)
        self.cpexec('<cpaneljson enable="1">')
       
    def cpexec(self, code, skip_return = False):
        if not self.connected:
            raise RuntimeError("The LiveAPI socket has closed, unable to continue")
        buffer = ""
        result = ""
        self.s.send(`len(code)` + "\n" + code)
       
       
        while True:
            buffer += self.s.recv(1024)
            pos = buffer.find("</cpanelresult>")
            if (pos != -1):
                break
        if skip_return:
            self._result = None
            return
        
        jsonstart = buffer.find("<cpanelresult>{")
        jsonend = buffer.find("</cpanelresult>")
        if jsonstart != -1 and jsonend != -1:
            jsonstart +=14
            parsed = None
            try:
                parsed = buffer[jsonstart:jsonend]
            except:
                parsed = None
            if buffer.find('<cpanelresult>{"cpanelresult"') == -1 and parsed is not None:
                jsonParsed = json.loads(parsed)
                self._result = {}
                self._result[u'cpanelresult'] = jsonParsed
            else:
                jsonParsed = json.loads(parsed)
                self._result = {}
                self._result = jsonParsed
        elif buffer.find("<cpanelresult></cpanelresult>") != -1:
             self._result = {'cpanelresult':{'error':'Error cannot be propagated to liveapi, please check the cPanel error_log', 'result':{'errors':['Error cannot be propagated to liveapi, please check the cPanel error_log.']}}}
        elif buffer.find("<cpanelresult>") != -1:
            #xml response un supported
            self._result = {'cpanelresult':{'error':'Error cannot be propagated to liveapi, please check the cPanel error_log', 'result':{'errors':['Error cannot be propagated to liveapi, please check the cPanel error_log.']}}}
            
        return self._result

    def api(self, reqtype, version, module, func, args = {}):
        if not self.connected:
            raise RuntimeError("The LiveAPI socket has closed, unable to continue.")
        input = {"module": module, "reqtype" : reqtype, "func": func, "apiversion":version}
        if len(args) > 0:
            input["args"] = args
        code = "<cpanelaction>\n" + json.dumps(input) + "\n</cpanelaction>"
        return self.cpexec(code);

    def api1(self, module, func, args = {}):
        if not self.connected:
            raise RuntimeError("The LiveAPI socket has closed, unable to continue.")
        return self.api("exec", "1", module, func, args)

    def api2(self, module, func, args = {}):
        if not self.connected:
            raise RuntimeError("The LiveAPI socket has closed, unable to continue.")
        return self.api("exec", "2", module, func, args)

    def api3(self, module, func, args = {}):
        if not self.connected:
            raise RuntimeError("The LiveAPI socket has closed, unable to continue.")
        return self.api("exec", "3", module, func, args)

    def uapi(self, module, func, args = {}):
        if not self.connected:
            raise RuntimeError("The LiveAPI socket has closed, unable to continue.")
        return self.api("exec", "uapi", module, func, args)

    def fetch(self, var):
        if not self.connected:
            raise RuntimeError("The LiveAPI socket has closed, unable to continue.")
        return self.cpexec('<cpanel print="' + var + '">')

    def cpanelif(self, code):
        if not self.connected:
            raise RuntimeError("The LiveAPI socket has closed, unable to continue.")
        return self.cpexec('<cpanel print="' + code + '">')

    def cpanelfeature(self,feature):
        if not self.connected:
            raise RuntimeError("The LiveAPI socket has closed, unable to continue.")
        return self.simple_result(self.api('feature','','feature', 'feature', feature))

    def cpanelprint(self,var):
        if not self.connected:
            raise RuntimeError("The LiveAPI socket has closed, unable to continue.")
        return self.simple_result(self.api1('print','',var))

    def cpanellangprint(self, key):
        if not self.connected:
            raise RuntimeError("The LiveAPI socket has closed, unable to continue.")
        return self.simple_result(self.api1('langprint','',key))

    def end(self):
        if not self.connected:
            raise RuntimeError("The LiveAPI socket has closed, unable to continue.")
        self.cpexec('<cpanelxml shutdown="1" />', True)
        self.s.close()

   
    def simple_result(self, result_data):
        return result_data['cpanelresult']['data']['result']